fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/DownloadTaskRequest.kt

```diff
@@ -1,28 +1,36 @@
 package com.github.kittinunf.fuel.core.requests
 
+import com.github.kittinunf.fuel.core.DefaultBody
+import com.github.kittinunf.fuel.core.ProgressCallback
 import com.github.kittinunf.fuel.core.Request
 import com.github.kittinunf.fuel.core.Response
 import com.github.kittinunf.fuel.util.copyTo
 import java.io.File
+import java.io.FileInputStream
 import java.io.FileOutputStream
 import java.net.URL
 
 internal class DownloadTaskRequest(request: Request) : TaskRequest(request) {
-    var progressCallback: ((Long, Long) -> Unit)? = null
+    var progressCallback: ProgressCallback? = null
     lateinit var destinationCallback: ((Response, URL) -> File)
 
     override fun call(): Response {
         val response = super.call()
         val file = destinationCallback(response, request.url)
+        var totalBytes = 0L
 
         FileOutputStream(file).use {
-            response.dataStream.copyTo(out = it, bufferSize = BUFFER_SIZE, progress = { readBytes ->
-                progressCallback?.invoke(readBytes, response.contentLength)
-            }) {
-                response.data = it
-            }
+            response.body.toStream().copyTo(out = it, bufferSize = BUFFER_SIZE, progress = { readBytes ->
+                // The Content-Length can *ONLY* be used if it's present, which it is not if the response is chunked or
+                // encoded (e.g. gzipped). We can opt to not send progress reports, send -1 or act as if it's at 100%
+                // constantly.
+                totalBytes = if (response.contentLength > 0) { response.contentLength } else { readBytes }
```

Comment:
[Nits] We don't need braces i think? `if (response.contentLength > 0) response.contentLength else readBytes`

Reply:
Sure! Habit from other languages